include/render/vulkan.h

```diff
@@ -0,0 +1,252 @@
+};
+
+// wlr_vk_render_surface implementing an own offscreen swapchain
+// might be created with a gbm device, allowing it to return a gbm_bo
+// for the current front buffer
+struct wlr_vk_offscreen_render_surface {
+	struct wlr_vk_render_surface vk_rs;
+	struct gbm_device *gbm_dev;
+	uint32_t flags;
+
+	struct wlr_vk_offscreen_buffer {
+		struct gbm_bo *bo; // optional
+		VkDeviceMemory memory;
+		struct wlr_vk_swapchain_buffer buffer;
+	} buffers[3];
+
+	// we have 3 different buffers:
+	// - one to render to that is currently not read/presented (back)
+	// - one on which rendering has finished and that was swapped
+	//   and is formally the front buffer
+	// - the buffer that was the front buffer the last time we
+	//   swapped buffers (before front) but might still be read,
+	//   so cannot yet be used for rendering
+	struct wlr_vk_offscreen_buffer *back; // currently or soon rendered
+	struct wlr_vk_offscreen_buffer *front; // rendering finished, presenting
+	struct wlr_vk_offscreen_buffer *old_front; // old presented
+};
+
+// Initializes a wlr_vulkan state. This will require
+// the given extensions and fail if they cannot be found.
+// Will automatically try to find all extensions needed.
+// param debug: Whether to enable debug layers and create a debug callback.
+struct wlr_vulkan *wlr_vulkan_create(
+	unsigned ini_ext_count, const char **ini_exts,
+	unsigned dev_ext_count, const char **dev_exts,
+	bool debug);
+
+void wlr_vulkan_destroy(struct wlr_vulkan *vulkan);
+
+// Tries to find any memory bit for the given vulkan device that
+// supports the given flags and is set in req_bits (e.g. if memory
+// type 2 is ok, (req_bits & (1 << 2)) must not be null.
+// Set req_bits to 0xFFFFFFFF to allow all types.
+int wlr_vulkan_find_mem_type(struct wlr_vulkan *vulkan,
+	VkMemoryPropertyFlags flags, uint32_t req_bits);
+
+const enum wl_shm_format *get_vulkan_formats(size_t *len);
+const struct wlr_vk_pixel_format *get_vulkan_format_from_wl(
+	enum wl_shm_format fmt);
+
+struct wlr_vk_renderer *vulkan_get_renderer(struct wlr_renderer *wlr_renderer);
+struct wlr_vk_texture *vulkan_get_texture(struct wlr_texture *wlr_texture);
+struct wlr_vk_render_surface *vulkan_get_render_surface(
+	struct wlr_render_surface *wlr_rs);
+
+// render_surface api
+struct wlr_render_surface *vulkan_render_surface_create_headless(
+	struct wlr_renderer *renderer, uint32_t width, uint32_t height);
+struct wlr_render_surface *vulkan_render_surface_create_xcb(
+	struct wlr_renderer *renderer, uint32_t width, uint32_t height,
+	void *xcb_connection, uint32_t xcb_window);
+struct wlr_render_surface *vulkan_render_surface_create_wl(
+	struct wlr_renderer *renderer, uint32_t width, uint32_t height,
+	struct wl_display *disp, struct wl_surface *surf);
+struct wlr_render_surface *vulkan_render_surface_create_gbm(
+	struct wlr_renderer *renderer, uint32_t width, uint32_t height,
+	struct gbm_device *gbm_dev, uint32_t flags);
```

Comment:
I guess these are unimplemented?

Reply:
Implemented in render/vulkan/render_surface.c